dnf/transaction_sr.py

```diff
@@ -0,0 +1,359 @@
+                .format(major=major, major_supp=VERSION_MAJOR)
+        )
+
+
+def store_transaction(transaction, filename="transaction.json"):
+    data = {
+        "version": VERSION,
+    }
+    rpms = []
+    groups = []
+    environments = []
+
+    for tsi in transaction.packages():
+        if tsi.is_package():
+            rpms.append({
+                "action": tsi.action_name,
+                "nevra": tsi.nevra,
+                "reason": libdnf.transaction.TransactionItemReasonToString(tsi.reason)
+            })
+
+        elif tsi.is_group():
+            group = tsi.get_group()
+
+            group_data = {
+                "action": tsi.action_name,
+                "id": group.getGroupId(),
+            }
+
+            if tsi.action == libdnf.transaction.TransactionItemAction_INSTALL:
+                group_data["package_types"] = \
+                    libdnf.transaction.compsPackageTypeToString(group.getPackageTypes())
+
+            groups.append(group_data)
+
+        elif tsi.is_environment():
+            env = tsi.get_environment()
+
+            env_data = {
+                "action": tsi.action_name,
+                "id": env.getEnvironmentId(),
+            }
+
+            if tsi.action == libdnf.transaction.TransactionItemAction_INSTALL:
+                env_data["package_types"] = \
+                    libdnf.transaction.compsPackageTypeToString(env.getPackageTypes())
+
+            environments.append(env_data)
+
+    if rpms:
+        data["rpms"] = rpms
+
+    if groups:
+        data["groups"] = groups
+
+    if environments:
+        data["environments"] = environments
+
+    with open(filename, "w") as f:
+        json.dump(data, f, indent=4, sort_keys=True)
+        f.write("\n")
+
+    print(_("Transaction saved to {}.").format(filename))
+
+
+class TransactionReplay(object):
+    def __init__(self, base, fn):
+        self._base = base
+        self._filename = fn
+
+    def _verify_toplevel_json(self, replay_data):
+        fn = self._filename
+
+        if "version" not in replay_data:
+            raise TransactionFileError(fn, _('Missing key "{key}".'.format(key="version")))
+
+        if not isinstance(replay_data["version"], str):
+            raise TransactionFileError(fn, _('Unexpected type of "version", string expected.'))
+
+        check_version(replay_data["version"], fn)
+
+    def _replay_pkg_action(self, pkg_data):
+        try:
+            action = pkg_data["action"]
+            nevra = pkg_data["nevra"]
+            reason = pkg_data["reason"]
+        except KeyError as e:
+            raise TransactionError(
+                _('Missing object key "{key}" in an rpm.').format(key=e.args[0])
+            )
+
+        solution = dnf.subject.Subject(nevra).get_best_solution(
+            self._base.sack,
+            with_nevra=True,
+            with_provides=False,
+            with_filenames=False,
+            forms=[hawkey.FORM_NEVRA]
+        )
```

Comment:
It can be replaced by `self._base.sack.query.filterm(nevra_strict=nevra)`. It is faster than solution. Did I already propose that?